datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -733,6 +817,22 @@ impl Stream for SMJStream {
                         match self.current_ordering {
                             Ordering::Less | Ordering::Equal => {
                                 if !streamed_exhausted {
+                                    if self.filter.is_some()
+                                        && matches!(
+                                            self.join_type,
+                                            JoinType::Left | JoinType::LeftSemi
+                                        )
+                                    {
+                                        self.freeze_all()?;
+
+                                        if !self.output_record_batches.batches.is_empty()
```

Comment:
I wonder if adding some methods on `JoinedRecordBatches` would make this code easier to read - For example instead of directly accessing the fields, perhaps adding functions would allow 

```suggestion
                                        if !self.output_record_batches.is_empty()
```

Reply:
That totally makes sense, I feel it requires more methods to add(add/is_empty/clear), I'd prefer to do it as follow up, this PR is too large imho